src/signatures.adoc

```diff
@@ -0,0 +1,194 @@
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=signature]
+|===
+
+Each parameter in a signature has a name, expressed as a Ion identifier symbol.
+Restricting names to Java-style identifiers enables use of operator characters (like `?` and
+`{asterisk}`) for the syntax surrounding names, including qualified macro references.
+
+
+=== Parameter Shapes
+
+A macro’s “wire format”—the sequence of acceptable tokens in an E-expression—is determined by its
+parameters’ shapes.
+The shape of a parameter has two dimensions: its _base type_ and its grouping.
+The base type constrains the expression forms that can be used for each argument supplied to the
+parameter.
+Independently, a parameter is either _simple_, _grouped_, or a _rest parameter_; this dimension
+determines how the arguments supplied to the parameter are delimited within the overall invocation.
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=param-shape]
+|===
+
+
+=== Base Types
+
+The core of a parameter specification is its base type, which constrains the syntax of each
+argument (that is, the acceptable expression forms that can be used).
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=param-types]
+|===
+
+The _concrete types_ correspond to the usual Ion data types, from `null` and `bool` through
+`list` and `struct`.
+These have the obvious meanings, with the caveat that annotations are allowed, as are
+appropriately-typed _and untyped_ nulls.
+For example, the inputs `null.int` and `null.null` are acceptable to an `int`-typed argument,
+as are arbitrary annotations on either.
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=concrete-type]
+|===
+
+The _abstract types_ are select supertypes of the concrete types: `text` accepts both `symbol`
+and `string`; `number` accepts `int`, `decimal`, and `float`; `lob` accepts `blob` and `clob`;
+`sequence` accepts `list` and `sexp`; `any` accepts any value.
+Nulls and annotations are accepted as with the concrete types.
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=abstract-type]
+|===
+
+Collectively, the abstract and concrete types are the _tagged types_.
+Parameters of these types can use macro invocations in place of normal values.
+
+The _primitive types_ are subtypes of various concrete types that have particularly compact
+binary encodings.
+These include variable-length strings, symbols, signed ``int``s, unsigned ``uint``s,
+as well as fixed-width ``int``s, ``uint``s, and ``float``s, all of various widths between 8 and
+64 bits.
+These types are untagged, so they do not accept nulls, annotations, or macro invocations.
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=primitive-type]
+|===
+
+Finally, any visible macro can be used as a type, in which case the argument is written (in text)
+as an S-expression with elements matching that macro’s signature.
+As with tagless arguments, these arguments are serialized without any explicit indication of
+their type, since that’s implied by context.
+(Zero-parameter macros—constants—are acceptable but pointless.)
```

Comment:
Took me a second to understand what this means. It's not saying that zero-parameter macros themselves are pointless, but rather that including a zero-parameter macro in a macro signature is pointless. Might be worth clarifying.